Fix Bluebird Patch

arthurschreiber opened the pull request:
The previous patch was working _most_ of the time, but not always. Especially when returning a `thenable` - I was running into multiple cases where the CLS namespace was lost somewhere along the chain between "native" Bluebird promises and other, non-promise objects that implemented a `.then` method.

This changes the bluebird patch to wrap the `async` object that bluebird uses. This is based on the `process.domain` support in the bluebird code, and thus should be 100% correct.

arthurschreiber commented:
See https://github.com/petkaantonov/bluebird/blob/master/src/async.js#L32-L36 and https://github.com/petkaantonov/bluebird/blob/master/src/async.js#L44-L48 for the code to support domains inside bluebird.

TimBeyer commented:
Thanks for the PR. Let me check if this works correctly on our production system. Then I'll merge it.

arthurschreiber commented:
Thanks for the quick reply! :smile:

TimBeyer commented:
So I took a look, and with your code alone the tests (which I wish I had rewritten and not just copy pasted, they're a mess) fail on the most recent bluebird.
Can you check if it works for you to wrap both `invoke*` and `_addCallbacks`?
That way the tests still pass for me.

arthurschreiber commented:
I tried running the tests, but I don't have redis installed and thus was not able to get the results. Can you maybe paste the error output?

In our application, I'm actually using a slightly different patch, but it's basically doing the same as what I'm proposing here:

``` js
// Patch Bluebird's async handlers to bind to all CLS namespaces.
// This makes it play nicely with continuation-local-storage.
var async = require("bluebird/js/main/async");

var old_invokeLater = async.invokeLater;
async.invokeLater = function Async$invokeLater(fn, receiver, arg) {
  var namespaces = Object.keys(process.namespaces), i;
  for (i = 0; i < namespaces.length; i++) {
    fn = process.namespaces[namespaces[i]].bind(fn);
  }

  return old_invokeLater.call(this, fn, receiver, arg);
};

var old_invoke = async.invoke;
async.invoke = function Async$invoke(fn, receiver, arg) {
  var namespaces = Object.keys(process.namespaces), i;
  for (i = 0; i < namespaces.length; i++) {
    fn = process.namespaces[namespaces[i]].bind(fn);
  }

  return old_invoke.call(this, fn, receiver, arg);
};
```

I actually had 3 of our app's unit tests failing after I changed the bluebird patch from `_addCallbacks`:

``` js
    it("keeps the Registry context alive after executing the function", function() {
      Registry.set("outer-value", 42);

      var promise = Promise.resolve();

      Registry.run(function() {
        Registry.set("inner-value", 1234);

        promise = promise.then(function() {
          return this.adapter.transaction(function() { }).then(function() {
            assert.equal(42, Registry.get("outer-value"));
            assert.equal(1234, Registry.get("inner-value"));
          });
        }.bind(this));
      }.bind(this));

      return promise;
    });
```

(`Registry` is a CLS namespace, and each test case is wrapped within a `Registry.run` block).

The `assert.equal(1234, Registry.get("inner-value"));` line would fail, because `"inner-value"` would be undefined. I actually think this is the correct behaviour (the rest of the app works just fine), but I'm not sure, and that failure might be related to the redis failures you're seeing. The rest of our app works fine with the patch I propose here (~50k lines of code).

But maybe I'm misunderstanding the test failure here... What's your opinion on this, should `"inner-value"` be defined at this position or not?

TimBeyer commented:
The reason we needed the patch was basically that `cls` would lose its bindings after a promisified `redis` call. So external dependencies that might interact with native modules need to be explicitly bound. That's what the `_addCallbacks` thing fixes and the reason `redis` is used as a dependency for the tests.

TimBeyer commented:
Regarding the test, I'd say `inner-value` should be defined, but `outer-value` not, since anything outside of `run` should not affect the registry. But since you are resolving the promise that you chain on before the `run` call, I'm not really sure what the expected behavior should be like. It kind of makes sense that it only accesses the `outer-value` at that point because the promise chain has a different binding than the `.run` function.

arthurschreiber commented:
As I mentioned, every test run is also wrapped in a `run` call, so the `run` call inside the test case is already nested. And that's why I now think that actually both values should be available. Putting the existing patch plus the  `async` patches together, my test cases work as they did before, and `thenable`s retain the correct CLS context.

Let me quickly update this PR.

arthurschreiber committed:
Fix Bluebird Patch

The previous patch was working *most* of the time, but not always. Especially when returning a `thenable` - I was running into multiple cases where the CLS namespace was lost somewhere along the chain between `native` Bluebird promises and other, non-promise objects that implemented a `.then` method.

This changes the bluebird patch to wrap the `async` object that bluebird uses. This is based on the `process.domain` support in the bluebird code, and thus should be 100% correct.

arthurschreiber commented:
I pushed an updated version.

arthurschreiber commented:
Unfortunately, I don't have a real test case for the `thenable` issue I'm encountering. The problem happens if a `Promise#then` does not return a `Promise`, but an object that itself has a `then` method. Bluebird will then (asynchronously) "follow" this Promise by calling `.then` on it, and at that point the CLS context will be lost.

As the `async` functions of bluebird are binding the given callbacks to the current domain anyway, I think binding to the namespace at that point should be correct.

TimBeyer commented:
I'd say let's merge it since the tests still pass. 
Then we can still figure out what's going on with those `.then` functions.

arthurschreiber commented:
I think I actually found the root cause for my problems... :disappointed:

We're using `tedious` to talk to our database, and it uses `EventEmitters` a lot in internally, and I did not wrap the request callbacks explicitly using `ns.bind`. For whatever reason, that worked fine with `RSVP` and `continuation-local-storage`, but when we moved to `bluebird`, this started to break in some weird cases. I'm now binding the callbacks we pass down to `tedious` explicitly, and that seems to have resolved the problem.

(I also tried to use the `bindEmitter` functionality of `cls`, but it looks like I can't actually reach all the internal EventEmitters that `tedious` uses, and the cls context then gets lost anyway. But binding the callback function explicitly works fine.)

Anyway, thanks for the quick responses, and sorry for wasting your time!

arthurschreiber commented:
Actually, there might be more to this, and I think I might have a test case that shows the problem. Let me try to get this down to a minimal test case.

overlookmotel commented:
Hi @TimBeyer and @arthurschreiber.

This PR has been quiet for a while, but wanted to check if a resolution was ever found?

[Sequelize](http://sequelizejs.com/) uses CLS for passing database transactions and bluebird for promises, but they've taken the approach of shimming bluebird methods individually, as they found that cls-bluebird wasn't working for things like `Promise#each()` - but this feels to me not to be the best way to do it ideally.

My thinking is that it'd be much better if they (and everyone else) could use cls-bluebird, if it were possible to solve the problems. And it sounds like this PR got at least most of the way.

Just wondered if you were still maintaining this module and whether it's difficult to solve these problems?

TimBeyer commented:
Hi @overlookmotel,

we're still using this in production and never ran into any issues, but we also started explicitly binding some callbacks to the correct CLS context in some cases.

If there are things that are broken, it would be great if we could come up with some test cases for that. Then I could fix them.

arthurschreiber commented:
We use this +  some additional patches on the `bluebird/async` object to ensure that thenables also get treated correctly (this was needed when upgrading to bluebird 2.4.0+).

arthurschreiber commented:
That's what we use: https://gist.github.com/arthurschreiber/063f9c940ae786b02fb9

`Registry` should be either replaced by your cls namespace, or if you want to bind to all namespaces, you'll need to change the code accordingly.

overlookmotel commented:
@arthurschreiber Brilliant! Does that work in all cases? (including for the collection iteration methods like `.each()`?)

And does it work with latest bluebird?

arthurschreiber commented:
Yes to both your questions. :smile:

arthurschreiber commented:
Note: What I pasted in the gist is what we use in addition to what cls-bluebird is already patching.

overlookmotel commented:
@arthurschreiber Long time since you posted the gist, but one question...

I've been looking at the gist and trying to figure it out. I'm wondering why it's necessary to shim `async.throwLater`. I don't see any code in `throwLater` which adds something to the queues, and hence can't see where CLS context can get lost.

`async.throwLater` either calls `setTimeout` (which CLS has already shimmed to maintain context) or `this._schedule`. On node (unless you've set a custom scheduler) `this._schedule` calls either `setImmediate` or `process.nextTick` (which again CLS has already shimmed).

Are you using a custom scheduler which necessitates shimming? Or does this by any chance have to do with using the code in your shim in Atom/Electron where maybe `MutationObserver` is used for scheduling?

Sorry to bother you, but please let me know if you have time.

overlookmotel commented:
@arthurschreiber By the way, I've put together a repo based on your gist here: https://github.com/overlookmotel/cls-bluebird2

NB this is your gist code only, not including cls-bluebird.

And a few tests that run on different shims of bluebird for CLS, to test which shim does what: https://github.com/overlookmotel/cls-bluebird-test

arthurschreiber commented:
> Are you using a custom scheduler which necessitates shimming? Or does this by any chance have to do with using the code in your shim in Atom/Electron where maybe MutationObserver is used for scheduling?

I think neither of those is the case. I _really_ don't remember exactly the reason why we shimmed the methods we did, and unfortunately, I'm no longer with my previous employer so I can't simply go and check the test cases we had for this. It might very well be the case that we shimmed more than was necessary.

I think some of the shimming was also inspired by existing code inside `bluebird` that was supposed to bring compatibility with Node.JS `domain`s.

Sorry I can't be of more help, but I'm impressed by the amount of work you put into #6! 👍

overlookmotel commented:
Version 2.0.0 just released I think fixes all known issues. So closing this PR.
